### PR TITLE
Feature: 404 page

### LINKED ISSUE #940: [Feature]: Design a 404 page
### Description

The current 404 page is not very good looking and should direct users to the challenges page or supply some option to push them elsewhere that is interesting/

### Explanation

_No response_

### PR DESCRIPTION
**Pull request by falcon-head**
## Description

I've made some improvements to the 404 page. The changes cater to both logged-in and logged-out scenarios, providing clear options for users to navigate effectively. 

For logged-in users encountering a 404 page, two buttons are displayed
- Go to tracks: redirects the user to the tracks page
- I'm feeling lucky: redirects the user to a random challenge page which is fetched from the `Challenge` table

For logged-out users encountering a 404 page, one button is displayed
- Go to home: which redirects to the '/' home page

## Related Issue

Closes #940 

## How Has This Been Tested?

By navigating to an invalid URL

### Screenshots:

States when the user is not logged in and goes to an invalid URL

Web capture_23-10-2023_202631_localhost

localhost_3000_settingss(iPhone 12 Pro) (1)

States when the user is logged in and navigates to an invalid URL

Web capture_23-10-2023_20265_localhost

localhost_3000_settingss(iPhone 12 Pro)

### Video:

https://github.com/typehero/typehero/assets/32938743/8c6b521a-27a7-4692-aa0a-51643f6e0350

### CODE REVIEW

**Comment by falcon-head** [comment]
@bautistaaa i hope this is good ?

**Comment by bautistaaa** [comment]
why have different experiences for logged in and logged out? those pages are public

**Comment by falcon-head** [comment]
> why have different experiences for logged in and logged out? those pages are public

Ah! Forgot that they were public pages. Shall I remove the logged-in experience and are we okay with the design?

**Comment by bautistaaa** [comment]
cc @PickleNik @maxdemaio

**Comment by PickleNik** [comment]
Goodlookin to me

File: apps/web/src/app/[locale]/not-found.tsx [hunk 1]
```diff
@@ -1,8 +1,50 @@
...
+        <h1 className="font-mono text-9xl font-bold leading-tight md:text-[12rem]">404</h1>
+        <p className=" px-6 text-center font-mono text-base md:px-0 md:text-xl">
+          Looks Like You Took a Wrong Turn at the Interface. <br className="hidden md:inline" />
+          Let's Get You Back on Track
+        </p>
+      </div>
+      {randomChallengeSlug !== null ? (
+        <div className="flex gap-4 ">
+          <a href="/tracks">
+            <Button variant="default" size="lg">
```
Comment (bautistaaa) [inline apps/web/src/app/[locale]/not-found.tsx:25]: we dont have many tracks yet, send them to explore

File: apps/web/src/utils/server/get-random-challenge.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+'use server';
+import { prisma } from '@repo/db';
+import type { Session } from '@repo/auth/server';
+
+interface resultType {
+  slug: string;
+}
+
+export async function getRandomChallenge(session: Session | null) {
+  if (!session) {
```
Comment (bautistaaa) [inline apps/web/src/utils/server/get-random-challenge.ts:10]: dont need this check or the session anymore

File: apps/web/src/app/[locale]/not-found.tsx [hunk 3]
```diff
@@ -1,8 +1,50 @@
-export default function NotFound() {
+import { Button } from '@repo/ui/components/button';
+import { BackgroundGrid } from './_components/hero-illustration';
+import { getServerAuthSession } from '@repo/auth/server';
+import { getRandomChallenge } from '~/utils/server/get-random-challenge';
+
+export default async function NotFound() {
+  const session = await getServerAuthSession();
```
Comment (bautistaaa) [inline apps/web/src/app/[locale]/not-found.tsx:7]: can remove any deps on session

**Comment by bautistaaa** [comment]
small cleanup comments then we can merge

**Comment by falcon-head** [comment]
Updated 🚀

**Pull request merged**